### ISSUE TITLE
[HOLD for payment 2022-10-18] [$250] Unresponsive camera opens on iOS if the camera permission is denied - reported by @adeel0202

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8846]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. (Make sure you have denied the camera permission)
2. Go to settings >  profile
3. Tap on the avatar icon
4. Select Upload photo
5. Select Take photo
6. Observe that the camera opens and it's unresponsive

## Expected Result:
A popup should be shown that asks the user to enable permission (just like it is happening on Android)

## Actual Result:
The camera opens and doesn't respond

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- iOS

**Version Number:** 1.1.57-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/166191478-9cfdafcb-a67e-4ac3-966b-c8c0e6ba8c68.MP4

**Expensify/Expensify Issue URL:** 
**Issue reported by:**  @adeel0202
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1649245469303159

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

[this job](https://www.upwork.com/jobs/~01980182c5243607c0)

**Comment by conorpendergrast** [comment]
@adeel0202 do you have steps for 1 ("(Make sure you have denied the camera permission)")? I don't see that as an option in **Settings** > **New Dot**

**Comment by adeel0202** [comment]
@conorpendergrast, you can go to settings of your iPhone, select New Expensify app and disable the Camera permissions. Or, you can uninstall the app, install again, try to access the camera and when you will be asked to allow the camera permissions, just deny it.

**Comment by Julesssss** [comment]
I'm going to take this, I can reproduce it on iOS and I agree with the expected result.

**Comment by Julesssss** [comment]
Not overdue

**Comment by Julesssss** [comment]
Reverting to external as I won't be able to work on this for a while

**Comment by dharmik** [comment]
Proposal:
Once user denied permission of Camera & if user again click on it so at that time we know that user denied it so in this case we have only one option to redirect user to iOS default settings page and user have to manually enable Camera option and came back to app and click on that camera option to perform actions.

@mananjadhav @Julesssss

**Comment by mananjadhav** [comment]
Thanks @dharmik. I agree we should be showing redirection to settings as subsequent popup.

But could you please add some technical details to your proposal? Where are you going to make changes, which file? What change? Etc.

**Comment by dharmik** [comment]
Proposal @mananjadhav 

Update in File: src/components/AttachmentPicker/index.native.js

```diff
+import {PERMISSIONS, request, RESULTS} from 'react-native-permissions';
```

```diff
                 textTranslationKey: 'attachmentPicker.takePhoto',
-                pickAttachment: () => this.showImagePicker(launchCamera),
+                pickAttachment: () => this.showImagePicker(launchCamera, true),
```

```diff
-    showImagePicker(imagePickerFunc) {
+     showImagePicker(imagePickerFunc, isCamera = false) {
         return new Promise((resolve, reject) => {
-            imagePickerFunc(getImagePickerOptions(this.props.type), (response) => {
-                if (response.didCancel) {
-                    // When the user cancelled resolve with no attachment
-                    return resolve();
-                }
-                if (response.errorCode) {
-                    switch (response.errorCode) {
-                        case 'permission':
-                            this.showPermissionsAlert();
-                            break;
-                        default:
-                            this.showGeneralAlert();
-                            break;
+            if(isCamera && Platform.OS === 'ios') {
+                request(PERMISSIONS.IOS.CAMERA).then((result) => {
+                    if (result === RESULTS.GRANTED) {
+                        this.openImagePicker(imagePickerFunc, resolve, reject)
+                    } else if(result === RESULTS.BLOCKED){
+                        this.showPermissionsAlert()
                     }
+                  }); 
+            } else {
+                this.openImagePicker(imagePickerFunc, resolve, reject)
+            }
+        });
+     }
 
-                    return reject(new Error(`Error during attachment selection: ${response.errorMessage}`));
+     openImagePicker(imagePickerFunc, resolve, reject) {
+        imagePickerFunc(getImagePickerOptions(this.props.type), (response) => {
+            if (response.didCancel) {
+                // When the user cancelled resolve with no attachment
+                return resolve();
+            }
+            if (response.errorCode) {
+                if(response.errorCode === 'permission') {
+                    this.showPermissionsAlert();
+                } else {
+                    this.showGeneralAlert();
                 }
-
-                return resolve(response.assets);
-            });
+                return reject(new Error(`Error during attachment selection: ${response.errorMessage}`));
+            }
+            return resolve(response.assets);
         });
-    }
+     }
```

By the above changes these scenario will happen in iOS

When user select option Take a Photo

1)If user accepted the permission: It will redirect to the Camera

2)If user have denied the permission: It will show the alert to go to setting and enable the camera permission then after we can access the camera

You can also refer bellow screenshots regarding changes

When first time we click on **Take a Photo**
![app-camera-permission-alert](https://user-images.githubusercontent.com/1510175/170714576-2411f9d2-44b5-410c-b0d7-463ba71d81cb.PNG)

When camera permission is denied or blocked
![app-permission-setting](https://user-images.githubusercontent.com/1510175/170714599-fb2feca3-40b8-4ab3-8bc2-9dc47afc78dd.PNG)

For enable Camera permission while tap on Setting
![app-camera-permission](https://user-images.githubusercontent.com/1510175/170714596-9fa596ed-08d6-49a6-b84e-6da8b35a748b.PNG)

**Comment by mallenexpensify** [comment]
@mananjadhav can you please review @dharmik 's proposal above?
https://github.com/Expensify/App/issues/8846#issuecomment-1139662273

**Comment by jboniface** [comment]
@mallenexpensify did this switch to daily on purpose? saw this with another issue too

**Comment by mananjadhav** [comment]
@dharmik Sorry for the delay here. I went through your proposal and it looks good, except that we don't allow `if(platform === 'ios')` checks in our code. If it existed before, then when changing a piece of the code, we refactor those pieces too. [We keep the platform-specific changes in their own files](https://github.com/Expensify/App/#platform-specific-file-extensions). Could you refactor your solution to cater to this? Say have `index.ios.js` and `index.android.js` instead of single `index.native.js`.

Another thing I'd like to highlight is having two versions doesn't mean a replica of the code with just platform-specific changes. We could DRY that part of the code and ensure the quality remains intact.

Thank you for your time and patience with me on this one.

**Comment by mallenexpensify** [comment]
Replied to other issue but pinging here too, it did switch to `Daily` because a proposal was submitted, ie. the world proposal is the first word from a non-employee in a post 
<img width="381" alt="image" src="https://user-images.githubusercontent.com/22508304/171696335-0e5ed350-9091-49e5-bbe5-07d4df4d3194.png">

**Comment by mananjadhav** [comment]
@dharmik Did you get a chance to look at my comment [here](https://github.com/Expensify/App/issues/8846#issuecomment-1144161773)?

**Comment by dharmik** [comment]
Your comment:

Another thing I'd like to highlight is having two versions doesn't mean a replica of the code with just platform-specific changes. We could DRY that part of the code and ensure the quality remains intact.


The issue is only from ios as issue says so if i create seperate file for ios and android so that could be more repetitive code.

I apology it confuse me. Can you please share your opion on it.

**Comment by Julesssss** [comment]
> The issue is only from ios as issue says so if i create seperate file for ios and android so that could be more repetitive code.

Manan was correct, his [link](https://github.com/Expensify/App/#platform-specific-file-extensions) shows the correct pattern for implementing fixes for specific platforms.

**Comment by mananjadhav** [comment]
@dharmik A basic example for the platform-specific files looks like this:

```
   test.js
   let variableX =  Platform.OS === 'ios'  ? 10: 20; // Assume noop for web and for iOS we need 10 and Android we need 20.

```

will become  
```
test/
   index.js
        import getVariableX from './getVariableX';

        let variableX = getVariableX();

   getVariableX/
         index.js
            export () => {}
         index.io.js
            export () => { return 10; }
         index.android.js
            export () => { return 20; }
  
```

Do let me know if you have any further questions.

**Comment by mananjadhav** [comment]
@dharmik Do you have any updates for us?

**Comment by dharmik** [comment]
Will check it today.

**Comment by mananjadhav** [comment]
Hi @dharmik, checking if you have an ETA for us on this one?

**Comment by dharmik** [comment]
@mananjadhav Below must be iOS specific file that you mean?
```
if(isCamera && Platform.OS === 'ios') {
+                request(PERMISSIONS.IOS.CAMERA).then((result) => {
+                    if (result === RESULTS.GRANTED) {
+                        this.openImagePicker(imagePickerFunc, resolve, reject)
+                    } else if(result === RESULTS.BLOCKED){
+                        this.showPermissionsAlert()
                     }
+                  }); 
+            }
```

**Comment by mananjadhav** [comment]
Yes @dharmik thats right. Anything with `Platform.OS` will get moved to its platform specific file.

**Comment by mananjadhav** [comment]
Thanks for the updated proposal @dharmik . Appreciate you sticking around for it. @Julesssss, @dharmik's updated proposal looks good.

🎀 👀 🎀
C+ reviewed

**Comment by mallenexpensify** [comment]
Created new job https://www.upwork.com/jobs/~01c9e1e61d8453232e
Hired @adeel0202 for reporting, @mananjadhav for C+ and @dharmik for the fix. All $250

**Comment by mananjadhav** [comment]
@dharmik Do you have any PR for me to review?

**Comment by arielgreen** [comment]
bump @dharmik

**Comment by dharmik** [comment]
For Information: I am facing compilation issue so I just resolved it so I am planing to raise PR in 1-2 days.

**Comment by mananjadhav** [comment]
Thanks for the comment @dharmik, will wait for it.

**Comment by dharmik** [comment]
@mananjadhav Just for update, I am able to compile it. Working on PR.

**Comment by Julesssss** [comment]
PR is being reviewed currently

**Comment by arielgreen** [comment]
@mananjadhav can you provide an update on the review please?

**Comment by mananjadhav** [comment]
Sorry @arielgreen I thought the failed checks will be picked up. I've posted the comment here https://github.com/Expensify/App/pull/9949#issuecomment-1209782992

**Comment by Julesssss** [comment]
The commits need to be signed before we can review and merge.

**Comment by dharmik** [comment]
@Julesssss @mananjadhav PR link https://github.com/Expensify/App/pull/10348

**Comment by Julesssss** [comment]
Hi @dharmik, there are some unresolved comments. Please take a look and let us know if you have further questions: https://github.com/Expensify/App/pull/10348#issuecomment-1221467227

**Comment by arielgreen** [comment]
@dharmik Please provide an update.

**Comment by mananjadhav** [comment]
@dharmik We've sent out multiple reminders. Please send an update within a day, else we'll have to rescind the job.

cc - @arielgreen @Julesssss

**Comment by arielgreen** [comment]
https://expensify.slack.com/archives/C02NK2DQWUX/p1662468554657719

**Comment by arielgreen** [comment]
Reposted job: https://www.upwork.com/jobs/~018cf2c599b90afaee

**Comment by hoangndm** [comment]
**Proposal**

I think you should check open camera permission there, if yes then open the camera, if not you should show a popup asking the user to open permission and show the button go to setting.

**Comment by Julesssss** [comment]
Thanks, @hoangndm, but we'll need a proposal on this issue. Something like this: https://github.com/Expensify/App/issues/8846#issuecomment-1139662273

**Comment by Julesssss** [comment]
Is there a reason that we can't use a native mobile only js file, instead of having similar solutions for both iOS & Android separately?

**Comment by Julesssss** [comment]
We prefer to not use platform conditionals, so it sounds like we'll need two files then :/

**Comment by mananjadhav** [comment]
@aimane-chnaif Your proposal looks good to me. I am only concerned about the need for updating the `Podfile`, but it seems it loads the pop of settings the next time? If that's not the case any specific reason why we need to add it now? Nevertheless your code seems to work, so I am find to handle that in the PR.

🎀 👀 🎀 
C+ reviewed

**Comment by Julesssss** [comment]
I recently upgraded XCode and I'm having some serious problems with my XCode env and phsical device. I will get to this soon! 😫

**Comment by arielgreen** [comment]
@aimane-chnaif sent an offer in Upwork

**Comment by arielgreen** [comment]
@Julesssss is this waiting on something from you?

**Comment by Julesssss** [comment]
Yeah, thanks @aimane-chnaif. 

Also, the App deploy has been blocked so the automated 7 days... message wasn't posted to this issue yet. It should occur today hopefully 🤞

**Comment by arielgreen** [comment]
Paid.

**Comment by mananjadhav** [comment]
@arielgreen The C+ payment is pending for this one.
